File: Queue/C#/Queue.cs
namespace C_
{
    public class MyQueue<T>
    {
        public Node<T>? Start { get; set; }
        public Node<T>? End { get; set; }

        public MyQueue()
        {
            Start = null;
            End = null;
        }

        public Node<T> GetNode(T value)
        {
            return new Node<T>(value);
        }

        public bool IsEmpty()
        {
            return this.Start == null || this.End == null;
        }

        public void Insert(T value)
        {
            Node<T> node = new(value);

            if (IsEmpty())
            {
                Start = node;
                End = node;
                return;
            }

            this.End!.Next = node;
            this.End = node;
        }

        public void Remove()
        {
            if (!this.IsEmpty())
            {
                if (this.Start == this.End)
                {
                    this.Start = null;
                    this.End = null;
                    return;
                }

                this.Start = this.Start!.Next;
            }
        }

        public bool Query(T value, out Node<T>? currentNode, out Node<T>? previousNode)
        {
            currentNode = this.Start;
            previousNode = null;

            while (currentNode != null)
            {
                if (currentNode.Value!.Equals(value)) return true;
                previousNode = currentNode;
                currentNode = currentNode.Next;
            }

            currentNode = null;
            previousNode = null;

            return false;
        }

        public void Show()
        {
            if (IsEmpty())
            {
                Console.WriteLine("Queue is empty");
                return;
            }

            Node<T>? auxNode = this.Start!;

            Console.Write("Queue: [");
            while (auxNode != null)
            {
                var isLast = auxNode.Next == null;
                Console.Write(auxNode.Value + (isLast ? "" : " -> "));
                auxNode = auxNode.Next;
            }
            Console.WriteLine("]");
        }
    }
}

File: Queue/C++/src/Queue.cpp
#ifndef QUEUE_C
#define QUEUE_C

#include "Queue.h"
#include "Node.h"

template <typename T>
Queue<T>::~Queue() {
    Node<T>* current = start;

    while (current != nullptr) {
        Node<T>* next_node = current->next;
        delete current;
        current = next_node;
    }
}

template <typename T>
void Queue<T>::enqueue(const T value) {
    auto* new_node = new Node<T>(value);

    if (is_empty()) {
        start = new_node;
        end = new_node;
        return;
    }

    end->next = new_node;
    end = new_node;
}

template <typename T>
T Queue<T>::dequeue() {
    if (!is_empty()) {
        auto removed_node = start->value;

        if (start == end) {
            start = nullptr;
            end = nullptr;
        } else start = start->next;

        return removed_node;
    }

    return T{};
}

template <typename T>
bool Queue<T>::is_empty() {
    return start == nullptr || end == nullptr;
}

template <typename T>
bool Queue<T>::query(const T value, Node<T>* currentNode, Node<T>* previousNode) {
    currentNode = start;

    while (currentNode != nullptr) {
        if (currentNode->value == value) {
            return true;
        }

        previousNode = currentNode;
        currentNode = currentNode->next;
    }

    return false;
}

template <typename T>
T* Queue<T>::toArray(size_t& size) const {
    size = 0;
    Node<T>* current = top;

    while (current != nullptr)
    {
        size++;
        current = current->next;
    }

    T* values = new T[size];
    current = top;
    int index = 0;

    while (current != nullptr)
    {
        values[index] = current->value;
        index++;
        current = current->next;
    }

    return values;
}

#endif
